### ISSUE TITLE
[Enhancement]: Enable EPUB 3 Support in Audiobookshelf for Synced Audiobooks and Ebooks

### ISSUE THREAD

**Issue by ashwinm4friends** [advplyr/audiobookshelf#3084]
### Type of Enhancement

Server Backend

### Describe the Feature/Enhancement

Audiobookshelf currently supports various audiobook and ebook formats. This feature request is to enable support for EPUB 3 files, which allows for synchronized audiobooks and ebooks within a single file. EPUB 3 is a versatile and widely adopted format that can integrate both text and audio, providing a seamless reading and listening experience.

### Why would this be helpful?

Integrating EPUB 3 support in Audiobookshelf would provide several benefits:

	1.	Unified Experience: Users can enjoy a synchronized reading and listening experience without needing to switch between separate files for text and audio.
	2.	Enhanced Accessibility: EPUB 3 format supports features such as text highlighting and navigation, which can be beneficial for users with visual or reading impairments.
	3.	Improved Comprehension: Synchronized text and audio help in better comprehension and retention of information by reinforcing the content through both reading and listening.
	4.	Support for Dyslexic and Visually Impaired Users: The synchronization of audio and text can significantly aid users with dyslexia or visual impairments by allowing them to follow along with the text as they listen, improving their reading experience and accessibility.
	5.	User Convenience: This feature simplifies the process of managing and accessing audiobooks and ebooks, as users would only need to handle a single file format.
	
	For reference, Storyteller (https://smoores.gitlab.io/storyteller/) already provides the capability to create EPUB 3 files with synchronized audio and text.  By implementing EPUB 3 support, Audiobookshelf can leverage the advanced features of this format to offer a richer, more accessible, and more integrated user experience.
	
	
	
	
	
	

### Future Implementation (Screenshot)

The EPUB 3 format allows for precise synchronization of audio with text, enabling features such as word-level or sentence-level highlighting. As the audio plays, the corresponding text is highlighted, providing a dynamic and engaging reading experience. Implementing this feature in Audiobookshelf will bring significant value to users by offering a more interactive and supportive reading environment.

### Audiobookshelf Server Version

latest

**Comment by advplyr** [comment]
This reads like ChatGPT.

We have discussed this a lot #189 #1723 #601

**Comment by yourmoonlight** [comment]
If Audiobookshelf could achieve the synchronization of ePub and audiobooks, it would be a very popular feature.

**Comment by Daemonias** [comment]
> This reads like ChatGPT.
> 
> We have discussed this a lot #189 #1723 #601

I think it's something entirely else.

It wouldn't require abb to generate anything.
In a epub3 the text and the Audiofiles are already there and synchronised. Abb would just need to support it, maybe give us a toggle to switch between reading and listening or have both at once.

**Comment by coandco** [comment]
I too would be very interested in EPUB3 support.

**Comment by Daemonias** [comment]
@advplyr any update on this?

**Comment by nichwall** [comment]
There has not been additional work on e-reader support since this was opened.

**Comment by kurumushi** [comment]
For reference if this ever gets looked at, this is the epub standard being asked for:

https://www.w3.org/submissions/epub-mediaoverlays/

Referenced here as well: https://www.w3.org/TR/epub-33/#sec-media-overlays

**Comment by dabigc** [comment]
For those that want this today, [Storyteller](https://smoores.gitlab.io/storyteller/) looks like it might be a good fit (granted, I believe it's in beta). The fact that it has an API server makes me wonder if it might one day be extensible for use in ABS without ABS needing to perform the generation, etc. If EPUB 3 supports this natively, it seems like Storyteller might be good to generate the EPUB 3 files and ABS might then only need to support the standard. Just throwing this out there as I prefer ABS and am sticking with it but for those that are eager, this might be a good solution.

**Comment by coandco** [comment]
Yes, that's the feature request for now -- just play existing synced-audio/text EPUB3 files, even if they're generated somewhere else.

**Comment by dunia01** [comment]
This would be great, really.

**Comment by kurumushi** [comment]
After looking at it I think this is much more involved than everyone may think. When you play an audiobook via audiobookshelf, it just asks the browser to play the media file. So in order for this to happen, you'd need either A. Browsers themselves to support epub3 synced. Or B. an entirely custom player interface.

**Comment by Astorsoft** [comment]
This was said above, this has nothing to do with #189, #1723 or #601, it isn't about generating a merge of an audiobook and epub like Storyteller does (and it does it great) or faking it through text to speech, it's about being able to read/listen the output of storyteller in Audiobookshelf. 

Not sure what the original requester had in mind, but the feature would make more sense on the client side (ie the iOS or Android apps) rather than the server side. FYI Storyteller does have a smartphone client to play those epub3 files, but there is some appeal is centralizing everything into one app like Audiobookshelf.

**Comment by eserero** [comment]
hmmm - interestingly I just found out that calibre ebook viewer can manage epubs that I created with storyteller, and it can do so also in the web client. it works in the desktop and my phone browser even have some limited offline capabilities. the browser reader is actually very good - very feature rich and can do quite a bit. Colibrio seems even nicer, but I think its not free. I wonder if there is a way to use the calibre reader in audiobookshelf

**Comment by Jdiesel87** [comment]
I just got around to generating a few Epub3 books using Storyteller. The process was relatively easy. It would be great to be able to play these books in ABS.

**Comment by wallaby-gopher** [comment]
> hmmm - interestingly I just found out that calibre ebook viewer can manage epubs that I created with storyteller, and it can do so also in the web client. it works in the desktop and my phone browser even have some limited offline capabilities. the browser reader is actually very good - very feature rich and can do quite a bit. Colibrio seems even nicer, but I think its not free. I wonder if there is a way to use the calibre reader in audiobookshelf

were you able to sync progress from calibre to audiobookshelf? It'd be great to have progress the same across platforms.

**Comment by xrishox** [comment]
this would be a huge boon of a feature. it's also an obvious feature for a project that puts such a strong focus on supporting both ebooks and audio books. it wouldnt need to implement support to produce these types of books, since that is already handled by storyteller. it would just need to be able to read them.

**Comment by andersonimes** [comment]
One stepping stone here that would be nice for users of Storyteller might be to simply have the ability to play epub3 files like audio files - effectively treating them like m4b files today. This way, we at least wouldn't have to duplicate files to support both formats.

**Comment by flboy1758** [comment]
I installed Storyteller and was amazed how easy it was to create audio synced epubs. I would love to be able to use audiobook shelf to consume these files instead of needing to manage two different apps.

**Comment by mattague** [comment]
> One stepping stone here that would be nice for users of Storyteller might be to simply have the ability to play epub3 files like audio files - effectively treating them like m4b files today. This way, we at least wouldn't have to duplicate files to support both formats.

I feel like this would be a good alternative of doing the syncing in ABS is just too difficult.

**Comment by mattague** [comment]
> Stupid me, even though I use the mobile app, most of my comments above about web browsers doesn't apply there at all.
> 
> For reference then, it looks like the mobile clients are using [epub.js](https://github.com/futurepress/epub.js), so it's probably best to see if that supports synced audiobooks, and request changes over there before anything could be done in the audio bookshelf mobile app.

I don't know how relevant this still is (if ABS still uses that project), but that repo hasn't been updated in almost 6 years. I did find mention of a fork that has been worked on quite a bit:

https://github.com/intity/epub-js

**Comment by nichwall** [comment]
> I don't know how relevant this still is (if ABS still uses that project), but that repo hasn't been updated in almost 6 years. I did find mention of a fork that has been worked on quite a bit:
> 
> https://github.com/intity/epub-js

Yes, this is still the library being used and we know it's out of date, but haven't found and implemented a suitable replacement yet. An updated fork might be easier to switch to.

**Comment by coandco** [comment]
Might be worth checking out https://github.com/johnfactotum/foliate-js -- BookLore just switched to it for ebook reading and it's been good so far.

**Comment by Jdiesel87** [comment]
There is also Readium

https://github.com/readium

**Comment by Alexchii** [comment]
Definitely looking forward to this functionality!

**Comment by flboy1758** [comment]
I found this plugin for koreader that supports playing audio while highlighting each sentence as it is read. https://github.com/stradichenko/audiobook.koplugin

I'm not talking about any of the transcribing features, I am not asking for transcription to be added to audiobookshelf. I am talking about how they handle epub files that were created in storyteller.

Crazy that a plugin for koreader is the best way to read ebooks via audiobookshelf

**Comment by mattague** [comment]
For anyone else looking into this and maybe alternatives, [Bookorbit](https://github.com/bookorbit/bookorbit) just released version 3.0.0 with full EPUB 3 read along support. It's still a bit early, and you still can't create them with Bookorbit, the functionality is pretty nice.

**Comment by lichenophile** [comment]
In the meantime, instead of shifting book library servers entirely, some of you might get what you need from the ABS client, Absorb. Not epub3 support, but effectively the same result. It can even live transcribe without an epub

https://github.com/pounat/absorb/issues/379#issuecomment-5687578874

**Comment by flboy1758** [comment]
@lichenophile transcription and syncing audiobooks are not "effectively the same result" I want to listen to an audiobook narrated by a person and have that audio kept in sync with an epub. I would never want to listen to a transcription

**Comment by lichenophile** [comment]
> I want to listen to an audiobook narrated by a person and have that audio kept in sync with an epub

Did I state somewhere that I was against this idea? I encourage you to re-read my comment and link...

**Comment by Vito0912** [comment]
Btw. the comment in that linked issue does not make any sense.

Epub3 does not need server side support lol.

The syncing of book and audio could require server side functions, but it could be done client side. But the playing of epub3 files that are already there (and what the issue asks) do not need any server side support.

Just another thing that apparently is ABS's fault...

**Comment by mattague** [comment]
Not ABSs fault, but how else could it work? If I want to use EPUB3 files, and ABS does not support it, but I find a client which does, I now have to keep a regular audiobook copy, and an EPUB3 copy, because ABS doesn't even recognize there is anything but text in the EPUB. So I can't even just play the audio from the EPUB in the web player.

**Comment by mattague** [comment]
> In the meantime, instead of shifting book library servers entirely, some of you might get what you need from the ABS client, Absorb. Not epub3 support, but effectively the same result. It can even live transcribe without an epub
> 
> https://github.com/pounat/absorb/issues/379#issuecomment-5687578874 
> 

I love Absorb, by far my favorite client. This comment didn't really do it justice for anyone who doesn't go read the link.

It's not using TTS to create a narration. It's offering two things. Using speech to text it can

- on the client device itself, run a speech model to match up the audiobook and ebook files. 
- Create text to read from an audiobook if it has no companion ebook

**Comment by kurumushi** [comment]
@Vito0912 
> Btw. the comment in that linked issue does not make any sense.
>Epub3 does not need server side support lol.

If you're playing ebooks through the web interface, then that web player (on the server) definitely needs to support epub3.

**Comment by Vito0912** [comment]
Epub3 per se is just the standard. It's often used for the Media Overlays it brings with it that does the synced playback. But this is just an epubfile. You can read it like an epub on ABS

ABSes clients (not the server) do not support the media overlay (so audio), yes. But that is because the client does not support it, not the server. I and many other third party apps already added support for this. The trade of is, that it is synced as book progress. But that is by design of epub 3 books. ABS stated iirc that audiobook and ebook sync for actual audio files and ebook files is out of scope. But in theory a client can set ebook and audiobook progres.

The statement that ABS needs to support epub3 in order for clients to use it is absolutely wrong and any client can use epub3 standards if the user uploaded epub3 files. 
I never said the ABS client does support ebup3 nor did the linked issue names that they want ABS official clients to have normal audio features for epub3 files. Clients (although if hacky) could implement audio only even with epub3, without extra APIs.

**Comment by kurumushi** [comment]
I am not talking about clients, just about the built in web interface and the player there. But I suppose that link was in regards to a client, so we're also talking about different things. I do believe however, that if ABS supported it on their own interface, clients would be more likely to follow.

**Comment by Vito0912** [comment]
Then I am not sure why you pinged me. I explicitly meant the comment made on the linked issue. But I also less meant you.

ABS clients sadly do not support Epub3, yeah.

Edit: after your edit: Yeah I fully would appreciate if ABS would support EPUB3. It's awesome. And with the new frontend/client it will be way easier to add. I use the same lib they use and it is very easy. So it's hopefully just a matter of time (as for the mobile apps idk). But hopefully.

**Comment by nichwall** [comment]
> I am not talking about clients, just about the built in web interface and the player there. But I suppose that link was in regards to a client, so we're also talking about different things. I do believe however, that if ABS supported it on their own interface, clients would be more likely to follow.

The web client is separate from the server. It is provided by the server, but is distinct from the server backend. The server already supports epub3, it's just an epub file.

This is a feature request for the web client to support epub3.
